### ISSUE TITLE
False-positive invalid-argument-type for indirect passing of types

### ISSUE THREAD

**Issue by gryznar** [astral-sh/ty#3210]
### Summary

```python
from dataclasses import dataclass
from enum import StrEnum

class Foo(StrEnum):
    FOO = "foo"

@dataclass
class Bar:
    foos: list[Foo]

def foo(val: list[str]): ...

def bar(val: Bar):
    foo([Foo.FOO]) # OK
    foo(bar.foos) # Argument to function `foo` is incorrect: Expected `list[str]`, found `list[Foo]`
```

### Version

ty 0.0.28 (927aad261 2026-04-02)

**Comment by AlexWaygood** [comment]
Thanks for the report! However, this is expected behavior. Because the inline list does not have a type annotation, ty is allowed to infer the list as having type `list[str]`, which satisfies the annotation for the `val` parameter of `bar`. But because `Bar.foos` is explicitly annotated, ty has no choice but to infer the type of `bar.foos` as `list[Foo]`. `list[Foo]` is unfortunately not a subtype of `list[str]` even though `Foo` is a subtype of `str` (see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics for why), so it's correct for ty to emit a diagnostic when an object of type `list[Foo]` is passed to a parameter expecting `list[str]`.

If your real-world `foo` function is first-party code, changing the `val` parameter so that it's annotated with `collections.abc.Sequence[str]` would solve this issue (`Sequence`, unlike `list`, is covariant, so `list[Foo]` _is_ a subtype of `Sequence[str]` even though it's not a subtype of `list[str]`)
